js_modules/dagster-ui/packages/ui-core/src/app/Permissions.tsx

```diff
@@ -111,53 +113,67 @@ type PermissionDisabledReasons = Record<keyof PermissionsMap, string>;
   permissions: PermissionBooleans;
   disabledReasons: PermissionDisabledReasons;
-  loading: boolean;
 };
 
-type PermissionsContextType = {
+export type PermissionsResult = {
+  error?: ApolloError;
   unscopedPermissions: PermissionsMap;
   locationPermissions: Record<string, PermissionsMap>;
-  loading: boolean;
-  // Raw unscoped permission data, for Cloud extraction
   rawUnscopedData: PermissionFragment[];
 };
 
+type PermissionsContextType = ReturnType<typeof wrapPromise<PermissionsResult>>;
+
 export const PermissionsContext = React.createContext<PermissionsContextType>({
-  unscopedPermissions: extractPermissions([]),
-  locationPermissions: {},
-  loading: true,
-  rawUnscopedData: [],
+  read() {
+    return {
+      error: undefined,
+      unscopedPermissions: extractPermissions([]),
+      rawUnscopedData: [],
+      locationPermissions: {},
+    };
+  },
 });
 
 export const PermissionsProvider = (props: {children: React.ReactNode}) => {
-  const {data, loading} = useQuery<PermissionsQuery, PermissionsQueryVariables>(PERMISSIONS_QUERY, {
-    fetchPolicy: 'cache-first', // Not expected to change after initial load.
-  });
+  const wrappedQuery = useWrappedQuery<
+    PermissionsQuery,
+    PermissionsQueryVariables,
+    PermissionsResult
+  >(
+    {
+      query: PERMISSIONS_QUERY,
+      fetchPolicy: 'cache-first', // Not expected to change after initial load.
+    },
+    async (result) => {
+      const {data, error} = result;
 
-  const value = React.useMemo(() => {
-    const unscopedPermissionsRaw = data?.unscopedPermissions || [];
-    const unscopedPermissions = extractPermissions(unscopedPermissionsRaw);
+      const unscopedPermissionsRaw = data?.unscopedPermissions || [];
+      const unscopedPermissions = extractPermissions(unscopedPermissionsRaw);
 
-    const locationEntries =
-      data?.workspaceOrError.__typename === 'Workspace'
-        ? data.workspaceOrError.locationEntries
-        : [];
+      const locationEntries =
+        data?.workspaceOrError.__typename === 'Workspace'
+          ? data.workspaceOrError.locationEntries
+          : [];
 
-    const locationPermissions: Record<string, PermissionsMap> = {};
-    locationEntries.forEach((locationEntry) => {
-      const {name, permissions} = locationEntry;
-      locationPermissions[name] = extractPermissions(permissions, unscopedPermissionsRaw);
-    });
+      const locationPermissions: Record<string, PermissionsMap> = {};
+      locationEntries.forEach((locationEntry) => {
+        const {name, permissions} = locationEntry;
+        locationPermissions[name] = extractPermissions(permissions, unscopedPermissionsRaw);
+      });
 
-    return {
-      unscopedPermissions,
-      locationPermissions,
-      loading,
-      rawUnscopedData: unscopedPermissionsRaw,
-    };
-  }, [data, loading]);
+      return {
+        error,
+        unscopedPermissions,
+        locationPermissions,
+        rawUnscopedData: unscopedPermissionsRaw,
+      };
```

Comment:
This all looks good to me!  Just from reading this it's not clear that the use of useWrapped has suspense implications, it sorta just looks like a stylistic change moving the data transform into the hook.  Maybe we call out in the naming that this significantly changes the loading semantics.